src/Pulsar.Client/Auth/AuthenticationOauth2.fs

```diff
@@ -0,0 +1,117 @@
+﻿module Pulsar.Client.Auth.Oauth2
+
+open System
+open System.Text.Json
+open System.Net.Http
+open System.Text.Json.Serialization
+open Pulsar.Client.Api
+open Pulsar.Client.Auth.Oauth2Token
+
+type Metadata =
+   {
+        [<JsonPropertyName("issuer")>]
+        Issuer:string
+        
+        [<JsonPropertyName("token_endpoint")>]
+        TokenEndpoint:string
+        
+        [<JsonPropertyName("userinfo_endpoint")>]
+        UserInfoEndpoint:string
+        
+        [<JsonPropertyName("revocation_endpoint")>]
+        RevocationEndpoint:string  
+               
+        [<JsonPropertyName("jwks_uri")>]
+        JwksUri:string
+        
+        [<JsonPropertyName("device_authorization_endpoint")>]
+        DeviceAuthorizationEndpoint:string 
+    }
+type Credentials =
+    {  
+        [<JsonPropertyName("type")>]
+        CredsType : string
+        
+        [<JsonPropertyName("client_id")>]
+        ClientId : string
+        
+        [<JsonPropertyName("client_secret")>]
+        ClientSecret : string 
+
+        [<JsonPropertyName("client_email")>]
+        ClientEmail : string 
+
+        [<JsonPropertyName("issuer_url")>]
+        IssuerUrl : string 
+    }
+
+//Gets a well-known metadata URL for the given OAuth issuer URL.
+//https://tools.ietf.org/id/draft-ietf-oauth-discovery-08.html#ASConfig
+let getWellKnownMetadataUrl (issuerUrl: Uri) : Uri =
+    Uri(issuerUrl.AbsoluteUri + ".well-known/openid-configuration")
+
+let getMetadata (issuerUrl: Uri) : Async<Metadata> =
+    async {
+        use client = new HttpClient()
+        let metadataDataUrl = getWellKnownMetadataUrl issuerUrl        
+        let! response = client.GetStringAsync metadataDataUrl |> Async.AwaitTask
+        return JsonSerializer.Deserialize<Metadata> response        
+    }
+
+let createClient issuerUrl =
+    let data = getMetadata issuerUrl |> Async.RunSynchronously  
+    TokenClient(Uri(data.TokenEndpoint))
+
+let openAndDeserializeCreds uri =
+    let text = System.IO.File.ReadAllText uri    
+    JsonSerializer.Deserialize<Credentials>(text)    
+
+type AuthenticationOauth2(issuerUrl: Uri, credentials: Uri, audience: Uri) =
+    inherit Authentication()
+    let tokenClient  = createClient issuerUrl
+    let credentials = openAndDeserializeCreds(credentials.LocalPath) 
+    let mutable token : Option<TokenResult * DateTime> = None   
```

Comment:
Can we avoid mutable variable?

Reply:
I guess if we use like list of tokens with length of 1. If list is emtpy, then it's time to fetch. If not - then proceed to check expiration. Can not think of something better.